File: 0875-koko-eating-bananas/0875-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: list[int], h: int) -> int:
         mins = 1 
         maxs = 0
         for pile in piles:
            maxs = max(maxs , pile)
         while(mins < maxs):
            mid = mins +(maxs-mins)//2

            if(self.caneat(piles,h,mid)):
                maxs = mid
            else:
                mins = mid+1
         return mins
         
    def caneat(self , piles:list[int] , h:int , speed:int) -> bool:
         hours = 0
         for pile in piles:
            hours +=math.ceil(pile/speed)
         return hours <= h

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int minspeed = 1;

        int maxspeed = 0;
        for(int pile : piles){
            maxspeed = Math.max(maxspeed,pile);
        }
        while(minspeed < maxspeed){
            int mid = minspeed +(maxspeed-minspeed)/2;

            if(caneatbananas(piles , h , mid)){
                maxspeed = mid;
            }
            else minspeed = mid+1;
        }
        return minspeed;
    }
    private boolean caneatbananas(int[] piles ,int h , int speed){
        int hours = 0;
        for(int pile:piles){
            hours += (int)Math.ceil((double)pile/speed);
        }
        return hours <= h;
    }
}